include/zephyr/drivers/cellular.h

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2023 Bjarki Arge Andreasen
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file drivers/cellular.h
+ * @brief Public cellular network API
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CELLULAR_H_
+#define ZEPHYR_INCLUDE_DRIVERS_CELLULAR_H_
+
+/**
+ * @brief Cellular interface
+ * @defgroup cellular_interface Cellular Interface
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#include <zephyr/types.h>
+#include <zephyr/device.h>
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define CELLULAR_MODEM_INFO_IMSI_SIZE 15
+#define CELLULAR_MODEM_INFO_IMEI_SIZE 16
+#define CELLULAR_MODEM_INFO_ICCID_SIZE 21
+#define CELLULAR_MODEM_INFO_MODEL_SIZE 16
+#define CELLULAR_MODEM_INFO_MANUFACTURER_SIZE 10
+
+/** Cellular access technologies */
+enum cellular_access_technology {
+	CELLULAR_ACCESS_TECHNOLOGY_GSM = 0,
+	CELLULAR_ACCESS_TECHNOLOGY_GPRS,
+	CELLULAR_ACCESS_TECHNOLOGY_UMTS,
+	CELLULAR_ACCESS_TECHNOLOGY_EDGE,
+	CELLULAR_ACCESS_TECHNOLOGY_LTE,
+	CELLULAR_ACCESS_TECHNOLOGY_LTE_CAT_M1,
+	CELLULAR_ACCESS_TECHNOLOGY_LTE_CAT_M2,
+	CELLULAR_ACCESS_TECHNOLOGY_NB_IOT,
+};
+
+/** Cellular network structure */
+struct cellular_network {
+	/** Cellular access technology */
+	enum cellular_access_technology technology;
+	/**
+	 * List of bands, as defined by the specified cellular access technology,
+	 * to enables. All supported bands are enabled if none are provided.
+	 */
+	uint16_t *bands;
+	/** Size of bands */
+	uint16_t size;
+};
+
+struct cellular_signal {
+	/* RSSI signal */
+	int16_t rssi;
+	/* Bit Error Rate */
+	uint8_t ber;
+};
+
+struct cellular_modem_info
+{
+	/* International Mobile Equipment Identity */
+	uint8_t imei[CELLULAR_MODEM_INFO_IMEI_SIZE];
+	/* GSM model as a string */
+	uint8_t model[CELLULAR_MODEM_INFO_MODEL_SIZE];
+	/* Manufacturer name as a string */
+	uint8_t manufacturer[CELLULAR_MODEM_INFO_MANUFACTURER_SIZE];
+	/* International Mobile Subscriber Identity */
+	uint8_t imsi[CELLULAR_MODEM_INFO_IMSI_SIZE];
+	/* Integrated Circuit Card Identification Number (SIM) */
+	uint8_t iccid[CELLULAR_MODEM_INFO_ICCID_SIZE];
```

Comment:
For a large list of const strings, I will suggest `struct` of `const char *`
```
struct cellular_modem_info {
        const char *imei;
        const char *model;
        ...
```
where the members point to `char` buffers initialized by the modem driver. Then when the modem has initialized the char buffers, we only need to copy pointers to the internal buffers
``` 
int cellular_get_modem_info(const struct device *dev, struct cellular_modem_info * modem_info);
{
        modem_info->imei = (const char *)data->imei;
}
```
This avoids a bunch of copying, and allows the driver to determine and manage the sizes of the members of `struct cellular_modem_info`

That said, I would prefer using a getter like `int cellular_get_modem_info(enum modem_info_type type, char *info, size_t size);` This is more flexible as we can add info to enum modem_info_type, without modifying the struct cellular_modem_info struct, and we can return `-ENODATA` if the modem does not provide said info.

Reply:
makes perfect sense